feat(lattice): Make lattice geometries differentiable and backend-agn…

Stellogic opened the pull request:
This commit introduces a major refactoring of the `Lattice` module to support automatic differentiation and backend-agnostic operations. The core motivation is to enable the optimization of lattice geometric parameters, such as the lattice constant, by minimizing physical observables like energy.

This is achieved by removing the direct dependency on NumPy for all geometric calculations and replacing them with operations from the abstract `tc.backend`.

Key Features & Changes:

- **Differentiable Geometry:** All internal calculations in the `Lattice` classes, including distance matrices and neighbor finding, are now based on differentiable backend tensor operations. This allows for gradient-based optimization of lattice parameters.

- **Backend Abstraction:** The `Lattice` module is now decoupled from any specific backend (JAX, TensorFlow, PyTorch), ensuring consistent behavior across all supported frameworks.

- **All-to-All Interactions:** Added `get_all_pairs` to `Lattice` and an `interaction_scope` parameter to `heisenberg_hamiltonian` to support models requiring interactions beyond nearest-neighbors, such as Rydberg or Lennard-Jones potentials.

- **New Optimization Example:** A new example, `lennard_jones_optimization.py`, demonstrates how to find the equilibrium lattice constant of a crystal by minimizing the Lennard-Jones potential.

Backend API Enhancements:

- The `ExtendedBackend` interface has been expanded with new abstract methods (`sort`, `meshgrid`, `where`, `equal`, etc.) to support the required geometric computations.
- All backends (`jax`, `numpy`, `pytorch`, `tensorflow`) have been updated to implement these new methods.

Testing:

- Added a new `TestDifferentiability` suite to `test_lattice.py` to verify gradient calculations.
- Enhanced tests for `CustomizeLattice` and data type consistency across backends.
- Added a new test for all-to-all interactions in the Heisenberg model.

There are many debug log for debugging. When the code is ready , we can delete them if necessary.
I thought this pr may finish my project

refraction-ray commented on examples/lennard_jones_optimization.py:
```diff
@@ -0,0 +1,116 @@
+    """
+    Calculate the total Lennard-Jones potential energy for a given logarithm of the lattice constant (log_a).
+    """
+    lattice_constant = jnp.exp(log_a)
```
better illustrate the example with lattice generated inside the function, otherwise it seems a static lattice can also do the task as long as the `base_distance_matrix` is given? In more complex case, when there are more lattice parameters, the "simply scaling the original distance matrix" paradigm will fail.

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -116,7 +139,10 @@ def get_coordinates(self, index: SiteIndex) -> Coordinates:
-        return self._coordinates[index]
+        assert self._coordinates is not None
+        coords = self._coordinates[index]
+        logger.debug(f"[DEBUG-LATTICE] get_coordinates for index {index}: {coords}")
```
no need for so much logger debug info, you can remove most of them

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -231,13 +282,18 @@ def get_neighbor_pairs(
         :rtype: List[Tuple[SiteIndex, SiteIndex]]
         """
 
+        logger.debug(
```
there are too much newly added debug info, please remove them for a small diff

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -475,69 +563,113 @@ def _identify_distance_shells(
+        displacements = backend.expand_dims(all_coords, 1) - backend.expand_dims(
+            all_coords, 0
         )
+        dist_matrix_sq = backend.sum(backend.power(displacements, 2), axis=-1)
```
``displacements**2``

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -1228,98 +1464,66 @@ def _build_lattice(self, *args: Any, **kwargs: Any) -> None:
-        :param tol: The numerical tolerance for distance
-            comparisons. Defaults to 1e-6.
-        :type tol: float, optional
+        This method leverages the generic `_build_neighbors_by_distance_matrix`
```
how the performance drop without KDTree? I think neighbor building can be on non differentiable, as they are discrete pairs and cannot be differentiated anyway? only the real valued distance matrix has to be differentiated.

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -1477,24 +1695,31 @@ def get_compatible_layers(bonds: List[Tuple[int, int]]) -> List[List[Tuple[int,
+    sorted_edges = sorted(list({(min(bond), max(bond)) for bond in bonds}))
 
     layers: List[List[Tuple[int, int]]] = []
+    unassigned_edges = set(sorted_edges)
```
what is the point of the renaming, which can induce more than 10 lines of diff with no diff at all

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1,10 +1,43 @@
+from __future__ import annotations
```
what is this for, I dont think use `__future__` package nowadays is a good idea

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1,10 +1,43 @@
+except ImportError:
+    torch = None
+
+if TYPE_CHECKING:
```
there is no mypy requirements for tests, so these type checking processing can all be removed

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -107,7 +141,7 @@ def test_input_validation_mismatched_lengths(self):
         with pytest.raises(
             ValueError,
-            match="Identifiers and coordinates lists must have the same length.",
+            match="The number of identifiers must match the number of coordinates.",
```
what is the point to change the text, which can only increase the diff. please keep the diff as minimal as possible

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1748,3 +2020,545 @@ def test_layering_on_edge_cases():
+}
+
+
+@pytest.fixture(scope="function")
```
dont use these fixtures, tc has paramterized fixtures for different backend, please refer to other tests files for the usage

refraction-ray reviewed (changes requested):
whenever possible, please keep the comment, name convention etc. exactly the same as the original one to minimize the diff, currently there are two many meaning less changes for debug log, variable name change etc. Keep the change as minimal as possible, dont let ai to rewrite the good parts with meaningless changes (like removing/adding comments, debug info, etc).

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -587,151 +719,232 @@ def __init__(
+            f"[DEBUG-LATTICE] Finished _build_lattice. Total sites: {self.num_sites}"
+        )
+
     def _get_distance_matrix_with_mic(self) -> Coordinates:
```
this function is not used as only the vectorized version is called? what about the performance difference between the two methods

Stellogic commented:
@refraction-ray 
Thanks for your review!
I have modified the code based on the comments.

Removed debug logs: There were many test failures after refactoring the code, and I was unsure of the cause, so I added a lot of debug logs. I will remember to remove them before creating a pull request next time.

Deleted unnecessary methods.

Modified the build_neighbors method in CustomizeLattice: When differentiation is needed, kdtree is not used (as I believe kdtree does not support differentiation). When differentiation is not required, kdtree is used to improve efficiency.

Added tests for new backend-independent methods.

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -759,20 +855,20 @@ class SquareLattice(TILattice):
-        basis_coords = np.array([[0.0, 0.0]])
+        # Define orthogonal lattice vectors for a square.
+        lattice_vectors = backend.convert_to_tensor(
+            [[lattice_constant, 0.0], [0.0, lattice_constant]]
```
if lattice constant is already a tensor, is this form of definition for lattice vectors work?

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -808,19 +904,21 @@ class HoneycombLattice(TILattice):
-        # Define the coordinates of the two basis sites (A and B)
-        basis_coords = a * np.array([[0.0, 0.0], [1.0, 0.0]])  # Site A  # Site B
+        # Define the two primitive lattice vectors for the underlying triangular Bravais lattice.
+        lattice_vectors = [
```
lattice vector is a list instead of a tensor? different from square lattice convention

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -808,19 +904,21 @@ class HoneycombLattice(TILattice):
-        basis_coords = a * np.array([[0.0, 0.0], [1.0, 0.0]])  # Site A  # Site B
+        # Define the two primitive lattice vectors for the underlying triangular Bravais lattice.
+        lattice_vectors = [
+            [a * 1.5, a * backend.sqrt(backend.convert_to_tensor(3.0)) / 2],
```
no need to compute sqrt(3) in backend, it is a float anyway, it cannot be jitted or differentiated

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1,12 +1,23 @@
+try:
+    import torch
+except ImportError:
+    torch = None
```
no need to do all these, as pytest fixture tfb, jaxb etc can automatically skip the test without installation of the corresponding packages

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1748,3 +1997,530 @@ def test_layering_on_edge_cases():
+        ),
+    ]
+
+    @pytest.mark.parametrize("backend_name", ["numpy", "jax", "tensorflow", "pytorch"])
```
use the backend fixture as in test_backends.py, also apply the backend fixture on all tests including the originally exisiting one to see their robustness across backends

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -1228,95 +1344,207 @@ def _build_lattice(self, *args: Any, **kwargs: Any) -> None:
         tol = kwargs.get("tol", 1e-6)
-        logger.info(f"Building neighbors for CustomizeLattice up to k={max_k}...")
+        use_kdtree = kwargs.get("use_kdtree", True)
+        force_differentiable = kwargs.get("force_differentiable", False)
```
what is the point to keep both kwargs for one choice?

refraction-ray commented on tensorcircuit/templates/lattice.py:
```diff
@@ -1228,95 +1344,207 @@ def _build_lattice(self, *args: Any, **kwargs: Any) -> None:
-        This implementation uses scipy.pdist for a memory-efficient
-        calculation of pairwise distances, which is then converted to a
-        full square matrix.
+    def _compute_distance_matrix(self) -> Coordinates:
```
this method is the same as abstractlattice? no need to rewrite here?

refraction-ray commented:
overall very nice work! I have left some more comments. and more improtantly, the test_lattice.py need some rewriting to use the same backend test fixture as how the test is defined in test_backends.py. and all tests in test_lattice.py should check with different backends, unless specific tests not valid in all backends

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1384,29 +1591,46 @@ def test_remove_sites_successfully(self, initial_lattice):
+    )
+    def test_remove_nonexistent_identifier_raises_error(self, backend, initial_lattice):
         """Tests that removing a non-existent site fails."""
+        tc.set_backend(backend)
```
delete and pls check all files to delete this line

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1578,25 +1911,44 @@ def test_tilattice_mixed_bc_distance_matrix_is_correct(self):
-            dimensionality=2,
-            identifiers=list(range(4)),
-            coordinates=[[0, 0], [1, 1], [0, 1], [1, 0]],
+    # 使用工厂函数而不是预先实例化对象，避免跨 backend 复用已缓存 _distance_matrix（numpy 数组）
```
english comments

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -1578,25 +1911,44 @@ def test_tilattice_mixed_bc_distance_matrix_is_correct(self):
         and positive off-diagonal elements.
         """
+        tc.set_backend(backend)
+        # 重新实例化 lattice，确保没有跨 backend 的缓存副作用
```
english comment

refraction-ray commented on tensorcircuit/backends/abstract_backend.py:
```diff
@@ -596,6 +596,82 @@ def argsort(self: Any, a: Tensor, axis: int = -1) -> Tensor:
+
+    def meshgrid(self: Any, *args: Any, **kwargs: Any) -> Any:
+        """
+         Return coordinate matrices from coordinate vectors.
```
still one more space on the above line

refraction-ray replied:
still one more space on the above line!

refraction-ray commented on examples/lennard_jones_optimization.py:
```diff
@@ -67,8 +67,6 @@ def calculate_potential(log_a, epsilon=0.5, sigma=1.0):
     history["a"].append(K.exp(log_a))
     history["energy"].append(energy)
 
-    # (Removed previously added blanket NaN guard per reviewer request to keep example minimal.)
```
delete

refraction-ray commented on examples/lennard_jones_optimization.py:
```diff
@@ -120,5 +118,3 @@ def calculate_potential(log_a, epsilon=0.5, sigma=1.0):
     plt.legend()
     plt.grid(True)
     plt.show()
-else:
```
why if is still kept?

refraction-ray commented on tests/test_lattice.py:
```diff
@@ -2098,6 +2090,54 @@ def test_layering_on_various_lattices(lattice_instance):
     _validate_layers(bonds, layers)
 
 
+# --- Regression tests for backend-scalar lattice constants (PR fix) ---
```
delete

Stellogic commented:
Thanks for your review!
Here are the changes:
Replaced numpy with a use of the stop_gradient method, falling back to the original tensor and making a copy if the method is not applicable.

Confirmed that convert_to_tensor always has a dtype under normal conditions, so there is no need to throw an exception; float32 or 64 is now selected directly.

Removed unnecessary if-statements.

refraction-ray commented on examples/lennard_jones_optimization.py:
```diff
@@ -0,0 +1,119 @@
+to optimize crystal structure. It finds the equilibrium lattice constant that minimizes
+the total Lennard-Jones potential energy of a 2D square lattice.
+
+The optimization showcases the key Task 3 capability: making lattice parameters
```
users have no idea what taks 3 is, please rephrase

refraction-ray commented:
LGTM, thanks for the nice contribution!

refraction-ray commented:
Now, I believe the only remaining thing is to add a jupyter tutorial here: https://github.com/tensorcircuit/tensorcircuit-ng/tree/master/docs/source/tutorials, to better and systematically illustrate the basic usage and functionalities of the newly added lattice module, you can include ingredients of previous tests and examples. The final output of the ipynb will be like https://tensorcircuit-ng.readthedocs.io/en/latest/tutorials/qubo_problem.html, which is easier for the users to learn about the new useful module.
